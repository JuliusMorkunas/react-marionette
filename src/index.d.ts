import * as React from 'react';
//import * as Marionette from 'backbone.marionette';

type MarionetteView = any;
type MarionetteViewClass = any;

interface OwnProps {
    view: MarionetteViewClass;
    viewOptions: any;
    onUpdateOptions?: (view: MarionetteView, viewOptions: any, nextViewOptions: any) => boolean;
    className?: string;
}

export class MarionetteComponent extends React.Component<OwnProps> {
}

export class ReactBehavior {
}

export class ReactView {
    constructor(options?: {
        attributes?: any[];
        className?: string;
        el?: any;
        tagName?: string;
        render: () => React.ReactNode,
        mountPoint: 'onShow'|'onRender'
    });
}
