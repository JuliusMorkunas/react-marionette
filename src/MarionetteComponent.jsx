/**
 * Developer: Stepan Burguchev
 * Date: 2/5/2017
 * Copyright: 2015-2017 ApprovalMax
 *       All Rights Reserved
 *
 * THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF ApprovalMax
 *       The copyright notice above does not evidence any
 *       actual or intended publication of such source code.
 */

import Marionette from 'backbone.marionette';
import React from 'react';
import PropTypes from 'prop-types';

class MarionetteComponent extends React.Component {
    constructor(props) {
        super(props);
        this._el = null;
        this._view = null;
        this._hostRegion = null;
        this._regionManager = null;
    }

    componentDidMount() {
        this._regionManager = new Marionette.View();
        this._hostRegion = this._regionManager.addRegion('hostRegion', {
            el: this._el
        });
        this._rebuildView();
    }

    componentWillReceiveProps(nextProps) {
        if (this.props.view !== nextProps.view) {
            throw new Error(
                '[MarionetteComponent] error: `props.view` cannot be changed after the initial render.'
            );
        }
    }

    componentDidUpdate(prevProps) {
        if (this.props.onUpdateOptions) {
            this.props.onUpdateOptions(this._view, prevProps.viewOptions, this.props.viewOptions);
        }
        if (this._view.shouldViewRebuild && this._view.shouldViewRebuild(this.props.viewOptions)) {
            this._rebuildView();
        }
    }

    componentWillUnmount() {
        this._regionManager.destroy();
    }

    getRegion() {
        return this._hostRegion;
    }

    _rebuildView() {
        if (!this.props.view) {
            return;
        }

        const View = this.props.view;
        this._view = new View(this.props.viewOptions);
        this._hostRegion.show(this._view);
    }

    render() {
        return <div className={this.props.className} ref={el => (this._el = el)} />;
    }
}

MarionetteComponent.propTypes = {
    view: PropTypes.func.isRequired,
    viewOptions: PropTypes.object,
    onUpdateOptions: PropTypes.func,
    className: PropTypes.string
};

export default MarionetteComponent;
