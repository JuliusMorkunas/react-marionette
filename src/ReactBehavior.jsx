/**
 * Developer: Stepan Burguchev
 * Date: 2/5/2017
 * Copyright: 2015-2017 ApprovalMax
 *       All Rights Reserved
 *
 * THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF ApprovalMax
 *       The copyright notice above does not evidence any
 *       actual or intended publication of such source code.
 */

import Marionette from 'backbone.marionette';
import React from 'react';
import ReactDOM from 'react-dom';

export default Marionette.Behavior.extend({
    initialize(options) {
        if (!options.containerEl) {
            throw new Error('Missing options: containerEl');
        }
        if (!options.containerEl) {
            throw new Error('Missing options: render');
        }
    },

    onShow() {
        ReactDOM.render(
            this.options.render(),
            this.__getContainerEl()
        );
    },

    onDestroy() {
        ReactDOM.unmountComponentAtNode(this.__getContainerEl());
    },

    __getContainerEl() {
        if (this.options.containerEl) {
            return this.$(this.options.containerEl)[0];
        }
        return this.el;
    }
});
