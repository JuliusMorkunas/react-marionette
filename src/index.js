export { default as MarionetteComponent } from './MarionetteComponent';
export { default as ReactBehavior } from './ReactBehavior';
export { default as ReactView } from './ReactView';
