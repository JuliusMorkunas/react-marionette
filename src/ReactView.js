/**
 * Developer: Stepan Burguchev
 * Date: 2/5/2017
 * Copyright: 2015-2017 ApprovalMax
 *       All Rights Reserved
 *
 * THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF ApprovalMax
 *       The copyright notice above does not evidence any
 *       actual or intended publication of such source code.
 */

import Marionette from 'backbone.marionette';
import ReactBehavior from './ReactBehavior';

export default Marionette.ItemView.extend({
    initialize(options) {},

    template: false,

    behaviors() {
        return {
            ReactBehavior: {
                behaviorClass: ReactBehavior,
                containerEl: null,
                reducer: this.options.reducer,
                store: this.options.store,
                render: this.options.render
            }
        };
    }
});
