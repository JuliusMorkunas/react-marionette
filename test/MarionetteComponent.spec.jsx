import React from 'react';
import Marionette from 'backbone.marionette';
import { mount } from 'enzyme';
import { MarionetteComponent } from '../src/index';

const SimpleView = Marionette.ItemView.extend({
    template({ name }) {
        return `Hello, ${name}`;
    },
    templateHelpers() {
        return {
            name: this.options.name
        };
    },
    className: 'sample-view'
});

describe('MarionetteComponent', () => {
    it('should render simple marionette view', () => {
        const component = mount(
            <div className="react-block">
                <MarionetteComponent
                    className="marionette-wrapper"
                    view={SimpleView}
                    viewOptions={{
                        name: 'Marionette'
                    }}
                />
            </div>
        );
        const marionetteElementHtml = component.getDOMNode().getElementsByClassName('marionette-wrapper')[0].innerHTML;

        expect(component).toMatchSnapshot();
        expect(marionetteElementHtml).toMatchSnapshot();
    });
});
