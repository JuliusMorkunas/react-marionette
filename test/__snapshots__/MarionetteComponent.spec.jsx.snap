// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarionetteComponent should not update view when options change (w/o update checkers) 1`] = `"<div class=\\"sample-view\\">Hello, Marionette</div>"`;

exports[`MarionetteComponent should render simple marionette view 1`] = `
<div
  className="react-block"
>
  <MarionetteComponent
    className="marionette-wrapper"
    view={[Function]}
    viewOptions={
      Object {
        "name": "Marionette",
      }
    }
  >
    <div
      className="marionette-wrapper"
    />
  </MarionetteComponent>
</div>
`;

exports[`MarionetteComponent should render simple marionette view 2`] = `"<div class=\\"sample-view\\">Hello, Marionette</div>"`;

exports[`MarionetteComponent should update view when options change (w/ onUpdateOptions) 1`] = `"<div class=\\"sample-view\\">Hello, Billy</div>"`;

exports[`MarionetteComponent should update view when options change (w/ shouldViewRebuild that does the update itself) 1`] = `"<div class=\\"sample-view\\">Hello, Billy</div>"`;

exports[`MarionetteComponent should update view when options change (w/ shouldViewRebuild that returns true) 1`] = `"<div class=\\"sample-view\\">Hello, Billy</div>"`;
